### ISSUE TITLE
Determine and Document the High-Level Design of `rgframework`

### ISSUE THREAD

**Issue by indiv0** [indiv0/colonize#47]
Currently `rgframework` is very unstable and poorly defined as I am still experimenting with it.

Once I've worked on it some more I should start planning out the final look of the API and determine how it should function. Then once that is complete, I should document the design of the framework.
# Resources
- AI
  - GOAP
    - [Goal-Oriented Action Planning](http://alumni.media.mit.edu/~jorkin/goap.html)
    - [Goal Oriented Action Planning for a Smarter AI](http://gamedevelopment.tutsplus.com/tutorials/goal-oriented-action-planning-for-a-smarter-ai--cms-20793)
- Art
  - Isometric
    - [Isometric Tutorial #1 by Spherical Ice](http://orig15.deviantart.net/50ec/f/2011/071/1/6/isometric_tutorial_by_tysonxxxnocture-d3bgjok.png)
    - [Isometric Pixel Art Tutorial](http://orig04.deviantart.net/d759/f/2009/258/8/7/isometric_background_tutorial_by_meninasuitcase.jpg)
    - [Roof with Irregular Iso Lines](http://pixeljoint.com/files/isoroof_tut_zi.png)
    - [Pixel Art Tutorial: Isometric Circles](http://pixeljoint.com/files/iso_circle_tut.gif)
    - [Pixelblink Presents a Pixel Art Tutorial - Part 1 - Making a Cube](http://pixeljoint.com/files/pixelbinktutorial001.gif)
- Async IO
  - [Herman J. Radtke III - My Basic Understanding of mio and Asynchronous IO](http://hermanradtke.com/2015/07/12/my-basic-understanding-of-mio-and-async-io.html)
  - [RustByExample - Threads](http://rustbyexample.com/std_misc/threads.html)
  - [RustByExample - Channels](http://rustbyexample.com/std_misc/channels.html)
  - [Rust Docs - `std::sync::mpsc](https://doc.rust-lang.org/std/sync/mpsc/)
  - [Asynchronous IO in Rust](https://blog.skcript.com/asynchronous-io-in-rust-36b623e7b965)
- Concurrency
  - [Intel - Don't Dread Threads - Part 1](http://uds.ak.o.brightcove.com/740838651001/740838651001_1127703220001_Dont-Dread-Threads-P1-V1.mp4)
  - [Rust book - Concurrency](https://doc.rust-lang.org/book/concurrency.html)
  - [StackOverflow - How does Dwarf Fortress keep track of so many entities without losing performance?](http://gamedev.stackexchange.com/questions/32813/how-does-dwarf-fortress-keep-track-of-so-many-entities-without-losing-performanc)
  - [Internals of a lightweight task scheduler](http://www.enkisoftware.com/devlogpost-20150905-1-Internals_of_a_lightweight_task_scheduler.html)
- Content Management
  - [StackOverflow - In XNA, how do I dynamically load parts of a large 2D world map?](http://gamedev.stackexchange.com/questions/610/in-xna-how-do-i-dynamically-load-parts-of-a-large-2d-world-map)
- ECS
  - [Shane Enishry - Misconceptions of Component-Based Entity Systems](https://shaneenishry.com/blog/2014/12/27/misconceptions-of-component-based-entity-systems/)
  - [HeroesGrave/ecs-rs](https://github.com/HeroesGrave/ecs-rs)
  - [slide-rs/specs - Parallel ECS](https://github.com/slide-rs/specs)
  - [lholden/entity_system](https://github.com/lholden/entity_system)
  - [RDBMS With Code In Systems](http://entity-systems-wiki.t-machine.org/rdbms-with-code-in-systems)
  - [TomBebbington/ecs.rs](https://github.com/TomBebbington/ecs.rs)
  - [Managing Decoupling Part 4 - The ID Lookup Table](http://bitsquid.blogspot.ca/2011/09/managing-decoupling-part-4-id-lookup.html)
  - [gamedev.net - How should systems pick which entities to process?](http://www.gamedev.net/topic/667290-how-should-systems-pick-which-entities-to-process-entity-component-systems/)
  - [rsaarelm.github.io - Roguelike game architecture in Rust 3 - The entity component system](http://rsaarelm.github.io/Roguelike-architecture-in-Rust-3/)
  - [StackOverflow - What is the role of “systems” in a component-based entity architecture?](http://gamedev.stackexchange.com/questions/31473/what-is-the-role-of-systems-in-a-component-based-entity-architecture)
  - [StackOverflow - Entity communication: Message queue vs Publish/Subscribe vs Signal/Slots](http://gamedev.stackexchange.com/questions/1119/entity-communication-message-queue-vs-publish-subscribe-vs-signal-slots/1449#1449)
- FRP
  - [YouTube - Why Functional Reactive Programming](https://www.youtube.com/watch?v=gaG3tIb3Lbk)
  - [Wikipedia - Functional reactive programming](https://en.wikipedia.org/wiki/Functional_reactive_programming)
  - Carboxyl
    - [blog.ebopp.de - FRP in Rust](http://blog.ebopp.de/blog/2015/01/26/frp-in-rust/)
    - [aepsil0n/carboxyl](https://github.com/aepsil0n/carboxyl)
    - [aepsil0n/carboxyl_window](https://github.com/aepsil0n/carboxyl_window)
- Game Engines
  - [C4 Engine - High Level Overview](http://i.imgur.com/81zIY.png)
  - [gamedev.net - Enginuity, Part I](http://www.gamedev.net/page/resources/_/technical/game-programming/enginuity-part-i-r1947)
  - [Amazon.com - Game Coding Complete, Third Edition](http://www.amazon.com/dp/1584506806/?tag=stackoverfl08-20)
  - [Game Engine Architecture Book](http://www.gameenginebook.com/)
  - [deWiTTERS Game Loop](http://www.koonsolo.com/news/dewitters-gameloop/)
  - [Create a game Engine: Part VIII Performance](http://in2gpu.com/2015/03/15/create-a-game-engine-part-viii-performance/)
- Data-Oriented Design
  - [Shane Enishry - Data-Oriented Design Matters](https://shaneenishry.com/blog/2015/03/26/data-oriented-design-matters/)
  - [DICE - Introduction to Data-Oriented Design](http://www.dice.se/wp-content/uploads/2014/12/Introduction_to_Data-Oriented_Design.pdf)
- Game Loop
  - [StackOverflow - Tips for writing the main game loop?](http://gamedev.stackexchange.com/questions/651/tips-for-writing-the-main-game-loop)
  - [StackOverflow - When should I use a fixed or variable time step?](http://gamedev.stackexchange.com/questions/1589/when-should-i-use-a-fixed-or-variable-time-step)
- State
  - [Game Programming Patterns - State - Pushdown Automata](http://gameprogrammingpatterns.com/state.html#pushdown-automata)
  - [Wikipedia - Finite-state machine](https://en.wikipedia.org/wiki/Finite-state_machine)
  - [Pushy Automata](https://apanatshka.github.io/compsci/2016/05/15/pushy-automata/)
- Graphics
  - [Nvidia - Alpha Blending: To Pre or Not To Pre](https://developer.nvidia.com/content/alpha-blending-pre-or-not-pre)
  - [Nvidia - Batch Batch Batch!](http://www.nvidia.com/docs/IO/8228/BatchBatchBatch.pdf)
  - [PistonDevelopers/gfx_voxel - `texture.rs`](https://github.com/PistonDevelopers/gfx_voxel/blob/master/src/texture.rs)
    - An example of texture atlasing
  - [Porter/Duff Compositing and Blend Modes](https://github.com/PistonDevelopers/graphics/issues/1026)
  - [Wikipedia - Multiple buffering - Double buffering](https://en.wikipedia.org/wiki/Multiple_buffering#Double_buffering_in_computer_graphics)
  - [Wikipedia - Multiple buffering - Page flipping](https://en.wikipedia.org/wiki/Multiple_buffering#Page_flipping)
  - [Wikipedia - Alpha compositing](https://en.wikipedia.org/wiki/Alpha_compositing)
  - [Wikipedia - Z-buffering](https://en.wikipedia.org/wiki/Z-buffering)
  - [0fps - Meshing in a Minecraft Game](https://0fps.net/2012/06/30/meshing-in-a-minecraft-game/)
  - [StackOverflow - What is deferred rendering?](http://gamedev.stackexchange.com/questions/74/what-is-deferred-rendering/184#184)
  - Isometric
    - [StackOverflow - Isometric View of "Voxel-Style" Terrain](http://gamedev.stackexchange.com/questions/84670/isometric-view-of-voxel-style-terrain)
  - sRGB
    - [Gamma error in picture scaling](http://www.4p8.com/eric.brasseur/gamma.html)
    - [The sRGB Learning Curve](https://gamedevdaily.io/the-srgb-learning-curve-773b7f68cf7a#.3goer5lcd)
  - PistonDevelopers/graphics
    - [PistonDevelopers/graphics - Draw calls performance bottleneck](https://github.com/PistonDevelopers/graphics/issues/1026)
  - Glium
    - [Glium API Docs](http://tomaka.github.io/glium/glium/index.html)
    - [tomaka - The glium library](https://medium.com/@tomaka/the-glium-library-5be149d87dc1#.gmvpvm33l)
    - [tomaka - glium-azdo](https://github.com/tomaka/glium-azdo)
  - OpenGL
    - [open.gl](https://open.gl/)
      - [open.gl - Depth Stencils](https://open.gl/depthstencils)
      - [open.gl - Texture objects and parameters](https://open.gl/textures)
    - [opengl-tutorial](http://www.opengl-tutorial.org/)
    - [learnopengl.com - Rendering Sprites](http://learnopengl.com/#!In-Practice/2D-Game/Rendering-Sprites)
    - [docs.gl - OpenGL API Documentation](http://docs.gl/)
    - [OpenGL Wiki](https://www.opengl.org/wiki/)
    - [StackOverflow - Drawing lots of tiles with OpenGL, the modern way](http://gamedev.stackexchange.com/questions/30362/drawing-lots-of-tiles-with-opengl-the-modern-way)
    - [StackOverflow - What does “immediate mode” mean in OpenGL?](http://stackoverflow.com/questions/6733934/what-does-immediate-mode-mean-in-opengl)
    - [OGLdev - Modern OpenGL Tutorials](http://ogldev.atspace.co.uk/index.html)
    - [OpenGL 4 Reference Pages](https://www.opengl.org/sdk/docs/man/)
    - [OpenGL From Zero To Hero](http://in2gpu.com/opengl-3/)
    - [Wikipedia - Vertex Buffer Object](https://en.wikipedia.org/wiki/Vertex_Buffer_Object)
    - [Reddit - Why so muchobsolete OpenGL?](https://www.reddit.com/r/opengl/comments/3pd4fw/why_so_much_obsolete_gl/)
    - [Approaching Zero Driver Overhead in OpenGL](http://gdcvault.com/play/1020791/)
    - GLSL
      - [GLSL Core Tutorial - Index](http://www.lighthouse3d.com/tutorials/glsl-tutorial/glsl-core-tutorial-index/)
- Misc
  - [Wikipedia - Interval tree](https://en.wikipedia.org/wiki/Interval_tree)
  - [Wikipedia - K-ary tree](https://en.wikipedia.org/wiki/K-ary_tree)
  - [Wikipedia - Memoization](https://en.wikipedia.org/wiki/Memoization)
  - [Wikipedia - Moore neighborhood](https://en.wikipedia.org/wiki/Moore_neighborhood)
  - [Wikipedia - Scene graph](https://en.wikipedia.org/wiki/Scene_graph)
- Profiling
  - [Profiling Rust applications on Linux](https://llogiq.github.io/2015/07/15/profiling.html)
  - [OpenGL - gDEBugger](https://www.opengl.org/sdk/tools/gDEBugger/)
- Terrain Generation
  - [nolithius.com - World Generation Breakdown](http://www.nolithius.com/game-development/world-generation-breakdown)
  - [Procedural Map Generation with Biomes](https://gamedev.stackexchange.com/questions/31241/random-map-generation/31245#31245)
  - [Eliminating Lag in procedurally Generated terrain?](https://www.reddit.com/r/gamedev/comments/1ao2fb/eliminating_lag_in_procedurally_generated_terrain/)
  - [Generating structures over several Chunks](https://www.reddit.com/r/gamedev/comments/22x7y0/generating_structures_over_several_chunks/)
  - [Torch - Generate CPU FlameGraphs based on DWARF Debug Info](https://github.com/mrhooray/torch)
  - Github - brendangregg/FlameGraph - Stack Trace Visualizer](https://github.com/brendangregg/FlameGraph)
- Voxel
  - [Lets Make a Voxel Engine](https://sites.google.com/site/letsmakeavoxelengine/)
  - [VoxelQuest - How Does It Work?](http://www.voxelquest.com/news/how-does-it-work)
  - [Volume Rendering 101](http://graphicsrunner.blogspot.ca/2009/01/volume-rendering-101.html)
# Examples
- Art
  - Isometric
    - [Gnomoria - Forum - Page 3](http://www.neogaf.com/forum/showthread.php?t=480647&page=3)
- ECS
  - [kvark/yasteroids - event.rs](https://github.com/kvark/yasteroids/blob/master/src/event.rs#L17)
- Graphics
  - `Glium`
    - [tomaka/glium - `examples/blitting.rs`](https://github.com/tomaka/glium/blob/2438a3f17bf98ec0733d9c8712d24fbe2b4c559d/examples/blitting.rs)
    - [tomaka/glium - `examples/sprites-batching.rs`](https://github.com/tomaka/glium/blob/2438a3f17bf98ec0733d9c8712d24fbe2b4c559d/examples/sprites-batching.rs)
      - **sprites batching**
    - [adudney/ECEN4003_FinalProject - VoxelTerrainGenerator](https://github.com/adudney/ECEN4003_FinalProject)
    - [Berling/rusty_gauntlet](https://github.com/Berling/rusty_gauntlet)
  - `Gfx`
    - [PistonDevelopers/gfx_voxel - `texture.rs`](https://github.com/PistonDevelopers/gfx_voxel/blob/master/src/texture.rs)
      - Texture atlasing
    - Shading
      - [PistonDevelopers/skeletal_animation - `skinned_renderer.rs`](https://github.com/PistonDevelopers/skeletal_animation/blob/master/src/skinned_renderer.rs#L86-L135)
        - **batched rendering**
      - [PistonDevelopers/hematite - `shader.rs`](https://github.com/PistonDevelopers/hematite/blob/master/src/shader.rs)
        - **voxel shader**
      - [yeliknewo/souls - `main.rs`](https://github.com/yeliknewo/souls/blob/d6c0b8eb0b9222ceaa2cb21006c776ab47ae5206/src/main.rs)
      - [gfx-rs/gfx - `examples/deferred`](https://github.com/gfx-rs/gfx/tree/2c00b52568e5e7da3df227d415eab9f55feba5a9/examples/deferred)
      - [gfx-rs/gfx - `examples/triangle`](https://github.com/gfx-rs/gfx/tree/2c00b52568e5e7da3df227d415eab9f55feba5a9/examples/triangle)
      - [aepsil0n/gfx-func - `examples/triangle.rs`](https://github.com/aepsil0n/gfx-func/blob/bb27fc48b65898b2dba561130f1a03f1595960dc/examples/triangle.rs)
        - **FRP shading**
  - `PistonWindow`
    - zmoshansky/rage - Scene graph
      - [`main.rs`](https://github.com/zmoshansky/rage/blob/master/src/main.rs#L121-L122)
      - [`scene_graph/mod.rs`](https://github.com/zmoshansky/rage/blob/master/src/scene_graph/mod.rs)
      - [`layout.rs`](https://github.com/zmoshansky/rage/blob/master/src/layout/mod.rs)
    - [DarkEld3r/rrthozopsi - `resource_manager.rs`](https://github.com/DarkEld3r/rrthozopsi/blob/eeb8258f532239c71bc8a005bf4d11a2583bcf15/src/resource_manager/mod.rs#L17)
      - **searching for the `assets` folder**
  - `GlGraphics`
    - [tedsta/reforge - space_gui.rs](https://github.com/tedsta/reforge/blob/1c73375c53bd7ebbfff2b35cf86c46ed7b745c82/src/space_gui.rs#L792-L892)
  - [`rsaarelm/calx`](https://github.com/rsaarelm/calx)
    - [rsaarelm/phage - `worldview.rs`](https://github.com/rsaarelm/phage/blob/master/src/worldview.rs)
    - [rsaarelm/phage - `tilecache.rs`](https://github.com/rsaarelm/phage/blob/master/src/tilecache.rs)
      - **how to cache tiles thread-locally**
  - Sdl2
    - [ArcadeRS 1.1: A simple window](http://jadpole.github.io/arcaders/arcaders-1-1)
- Networking
  - [AngryLawyer/SuperMegaTag](https://github.com/AngryLawyer/SuperMegaTag)
  - [aweinstock314/HackRPI_2014_Project](https://github.com/aweinstock314/HackRPI_2014_project)
  - [PistonDevelopers/hematite_server](https://github.com/PistonDevelopers/hematite_server)
  - [jaredonline - Rougelike ECS gist](https://gist.github.com/jaredonline/84237bb323f0d1ed2b60)
  - [Playform Dev Blog - High-Level Client/Server Design](http://playformdev.blogspot.ca/2015/02/high-level-clientserver-design.html)
  - [Playform Dev Blog - Update: Client/Server Refactoring](http://playformdev.blogspot.ca/2015/03/update-clientserver-refactoring.html)
# Issues
- [Glium - Errors with multi-draw](https://github.com/tomaka/glium/issues/1359)

**Comment by indiv0** [comment]
Related to https://github.com/indiv0/colonize/issues/31.
